### PR TITLE
fix for ProcMon CommandLine Parse Problem

### PR DESCRIPTION
**Pull request by kaganisildak**
Normally, ProcMon tries to read the commandline information of the new created Process as "Unicode" via the "CommandLine" address from OS JSON Profile.

    _EPROCESS->_PEB->ProcessParameters(_RTL_USER_PROCESS_PARAMETERS)->CommandLine
    addr_t cmdline_addr = user_process_parameters_addr + f->command_line;
uses this way

However, in the OS JSON Profile file, CommandLine looks like a single field.
Normally, ProcMon does not have a problem and prints the CommandLine information, but if the CommandLine of the Process is long data, ProcMon cannot read and process it.(related to libvmi)

An error related to this was reported
#821

CommandLine(_UNICODE_STRING) has 3 fields on Windows systems.

      0xAAAAA0    Length                          [unsigned short:Length]: 0x1882
      0xAAAAA2    MaximumLength                   [unsigned short:MaximumLength]: 0x1884
      0xAAAAA8    Buffer                         <UnicodeString Pointer to [0x4d2608] (Buffer)>

The first 2 bytes show Buffer's length.

This patch processes and displays the data with the length of the CommandLine.

Without patch
Before

With patch
After

### CODE REVIEW

**Comment by kaganisildak** [comment]
The root cause of the condition is related to libvmi.  
ProcMon uses the "drakvuf_read_unicode" function to read the address.  
When the function is called, the "windows_read_unicode_struct_pm" function compares the length  of Buffer and PageSize while reading the address.  

       if ( buffer_len > VMI_PS_4KB ) {
    
        dbprint(VMI_DEBUG_READ, "--%s: the length of %" PRIu16 " in the UNICODE_STRING at 0x%" PRIx64 " is excessive, bailing out.\n",
        
        __FUNCTION__, buffer_len, ctx->addr);
        
        return NULL;
    
    }


In short, you can deactivate the control here or increase the page size, to get more data about process during the monitor. (VMI_PS_16KB)

File: src/plugins/procmon/procmon.cpp [hunk 1]
```diff
@@ -145,6 +145,39 @@ struct process_visitor_ctx
...
+    uint16_t cmd_len = 0;
+    if(VMI_SUCCESS == vmi_read_16(vmi,&ctx2,&cmd_len))
+    {
+        ctx2.addr = cmdline_addr+8;
+        addr_t buffer_adr = 0;
+        if(VMI_SUCCESS == vmi_read_addr(vmi,&ctx2,&buffer_adr))
+        {
+            ctx2.addr = buffer_adr;
+            char *buf_ret;
+            buf_ret = (char*)malloc(cmd_len+1);
```
Comment (icedevml) [inline src/plugins/procmon/procmon.cpp]: The rest of `procmon` plugin is relying on `g_malloc0` instead of `malloc`. You could use `g_malloc0` so the whole plugin would depend on the same memory allocator.

File: src/plugins/procmon/procmon.cpp [hunk 2]
```diff
@@ -145,6 +145,39 @@ struct process_visitor_ctx
...
+            if(VMI_SUCCESS == vmi_read(vmi,&ctx2,cmd_len,buf_ret,NULL))
+            {
+                cmd = (char*)malloc(cmd_len+1);
```
Comment (icedevml) [inline src/plugins/procmon/procmon.cpp]: Same as previously.

File: src/plugins/procmon/procmon.cpp (hunk 1 above)
Comment (icedevml) [inline src/plugins/procmon/procmon.cpp]: The rest of the `procmon` plugin is relying on `g_malloc0` instead of `malloc`. You could use `g_malloc0` so the whole plugin would depend on the same memory allocator.

File: src/plugins/procmon/procmon.cpp (hunk 2 above)
Comment (icedevml) [inline src/plugins/procmon/procmon.cpp]: Same as previously.

File: src/plugins/procmon/procmon.cpp [hunk 3]
```diff
@@ -263,6 +297,7 @@ static void print_process_creation_result(
 
     g_free(cmdline);
     g_free(curdir);
+    free(cmd);
```
Comment (icedevml) [inline src/plugins/procmon/procmon.cpp]: As previously, `g_free`?

File: src/plugins/procmon/procmon.cpp [hunk 4]
```diff
@@ -145,6 +145,39 @@ struct process_visitor_ctx
 
 } // namespace
 
+static char* readcmdline(vmi_instance_t vmi, drakvuf_trap_info_t* info, addr_t addr)
```
Comment (icedevml) [inline src/plugins/procmon/procmon.cpp]: Maybe `read_command_line` would better match the naming convention? (The rest of functions is using names with underscores).

File: src/plugins/procmon/procmon.cpp [hunk 5]
```diff
@@ -145,6 +145,39 @@ struct process_visitor_ctx
...
+            {
+                    .translate_mechanism = VMI_TM_PROCESS_DTB,
+                    .dtb = info->regs->cr3,
+                    .addr = addr,
+            };
+    addr_t cmdline_addr = ctx2.addr;
+    uint16_t cmd_len = 0;
+    if(VMI_SUCCESS == vmi_read_16(vmi,&ctx2,&cmd_len))
+    {
+        ctx2.addr = cmdline_addr+8;
```
Comment (icedevml) [inline src/plugins/procmon/procmon.cpp]: Where `+8` came from? If it's an offset in some structure, you could try to find it out using Rekall profile. If for some reason it's not possible, it would be good to define `8` as a constant with meaningful name.

**Comment by icedevml** [comment]
@drakvuf-jenkins This is OK to test.

File: src/plugins/procmon/procmon.cpp (hunk 5 above)
Comment (kaganisildak) [reply src/plugins/procmon/procmon.cpp]: CommandLine is a data of type _UNICODE_STRING.
+8 was used to find the Buffer address.

```
  "_RTL_USER_PROCESS_PARAMETERS": [1024, {
   "CommandLine": [112, ["_UNICODE_STRING", {}]], 
   "ConsoleFlags": [24, ["unsigned long", {}]], 
   "ConsoleHandle": [16, ["Pointer", {
    "target": "Void"
   }]]
```

```
 "_UNICODE_STRING": [16, {
   "Buffer": [8, ["Pointer", {
    "target": "unsigned short"
   }]], 
   "Length": [0, ["unsigned short", {}]], 
   "MaximumLength": [2, ["unsigned short", {}]]
  }]
```

File: src/plugins/procmon/procmon.cpp [hunk 6]
```diff
@@ -158,16 +158,16 @@ static char* readcmdline(vmi_instance_t vmi, drakvuf_trap_info_t* info, addr_t a
...
     {
-        ctx2.addr = cmdline_addr+8;
+        ctx2.addr = cmdline_addr+8; // _UNICODE_STRING->Buffer
         addr_t buffer_adr = 0;
         if(VMI_SUCCESS == vmi_read_addr(vmi,&ctx2,&buffer_adr))
         {
             ctx2.addr = buffer_adr;
             char *buf_ret;
-            buf_ret = (char*)malloc(cmd_len+1);
+            buf_ret = (char*)g_malloc0(cmd_len+1);
```
Comment (tklengyel) [inline src/plugins/procmon/procmon.cpp]: `g_try_malloc0` and check for NULL

File: src/plugins/procmon/procmon.cpp [hunk 7]
```diff
@@ -158,16 +158,16 @@ static char* readcmdline(vmi_instance_t vmi, drakvuf_trap_info_t* info, addr_t a
...
             if(VMI_SUCCESS == vmi_read(vmi,&ctx2,cmd_len,buf_ret,NULL))
             {
-                cmd = (char*)malloc(cmd_len+1);
+                cmd = (char*)g_malloc0(cmd_len+1);
```
Comment (tklengyel) [inline src/plugins/procmon/procmon.cpp]: `g_try_malloc0` and check for NULL

File: src/plugins/procmon/procmon.cpp [hunk 8]
```diff
@@ -145,6 +145,41 @@ struct process_visitor_ctx
...
+        if(VMI_SUCCESS == vmi_read_addr(vmi,&ctx2,&buffer_adr))
+        {
+            ctx2.addr = buffer_adr;
+            char *buf_ret;
+            buf_ret = (char*)g_try_malloc0(cmd_len+1);
+            if (!buf_ret) return NULL;
+            if(VMI_SUCCESS == vmi_read(vmi,&ctx2,cmd_len,buf_ret,NULL))
+            {
+                cmd = (char*)g_try_malloc0(cmd_len+1);
+                if (!cmd) return NULL;
```
Comment (tklengyel) [inline src/plugins/procmon/procmon.cpp]: I think you are leaking `buf_ret` if you just return NULL here.

File: src/plugins/procmon/procmon.cpp [hunk 9]
```diff
@@ -145,6 +145,41 @@ struct process_visitor_ctx
...
+                int i;
+                for(i = 0;i<cmd_len;i++)
+                {
+                    strncat(cmd,&buf_ret[i],1);
+                }
+            }
+        }
+    }
+    return cmd;
```
Comment (tklengyel) [inline src/plugins/procmon/procmon.cpp]: `buf_ret` seems to be leaked on return here

File: src/plugins/procmon/procmon.cpp (hunk 8 above)
Comment (tklengyel) [reply src/plugins/procmon/procmon.cpp]: You marked it resolved but I don't see it being freed?

File: src/plugins/procmon/procmon.cpp [hunk 10]
```diff
@@ -174,8 +174,9 @@ static char* read_cmd_line(vmi_instance_t vmi, drakvuf_trap_info_t* info, addr_t
                 for(i = 0;i<cmd_len;i++)
                 {
                     strncat(cmd,&buf_ret[i],1);
-                }
+                }                
             }
+            g_free(buf_ret);
```
Comment (kaganisildak) [inline src/plugins/procmon/procmon.cpp]: @tklengyel

Comment (icedevml) [reply src/plugins/procmon/procmon.cpp]: If this will return `NULL`:
https://github.com/tklengyel/drakvuf/pull/833/files#diff-e1df45a792096383aa407daca2f1f8acR171

then the line you've mentioned will never be reached and thus `buf_ret` will leak.

File: src/plugins/procmon/procmon.cpp (hunk 8 above)
Comment (kaganisildak) [reply src/plugins/procmon/procmon.cpp]: Is it not appropriate to freed buf_ret here?
[#L179](https://github.com/tklengyel/drakvuf/blob/caa0487d633f101745315a06a4d9a1f39c8424bd/src/plugins/procmon/procmon.cpp#L179) 
@tklengyel

Comment (icedevml) [reply src/plugins/procmon/procmon.cpp]: If `buf_ret` allocation succeeds but `cmd` not, then `cmd` is `NULL` and you will exit the function here:

https://github.com/tklengyel/drakvuf/blob/caa0487d633f101745315a06a4d9a1f39c8424bd/src/plugins/procmon/procmon.cpp#L172

Thus, line 179 will not be reached and you leak the memory.

Comment (kaganisildak) [reply src/plugins/procmon/procmon.cpp]: Is it okay?
[d0df43e](https://github.com/tklengyel/drakvuf/pull/833/commits/d0df43ea7279ac9d28bdd9edf1c27f8aaa1e241b)

Comment (icedevml) [reply src/plugins/procmon/procmon.cpp]:
```cpp
                if (!cmd){
                    g_free(buf_ret);
                    g_free(cmd);
                    return NULL;
                }
```

You can remove `g_free(cmd)` - if you are inside that `if` then `cmd` is guaranteed to be `NULL`  - unallocated.
